src/lib.rs

```diff
@@ -0,0 +1,166 @@
+mod fetcher;
+mod interface;
+use alloy::{
+    providers::{ProviderBuilder, RootProvider},
+    transports::http::Http,
+};
+use log::LevelFilter;
+use log4rs::{
+    append::file::FileAppender,
+    config::{Appender, Root},
+    encode::pattern::PatternEncoder,
+    Config,
+};
+use reqwest::{Client, Url};
+use sled::Tree;
+use std::str::FromStr;
+use thiserror::Error;
+
+use self::{
+    fetcher::fetch_prices,
+    interface::{PriceData, Serializable},
+};
+
+#[derive(Clone)]
+pub struct Configuration {
+    pub fetch_interval_seconds: u64,
+    pub contracts: Vec<(String, String)>,
+    pub provider: RootProvider<Http<Client>>,
+}
+
+/// ## Cryptoprices
+///
+/// Cryptoprices is a lightweight Rust library that provides your Rust applications with a direct
+/// link to the latest cryptocurrency prices. All data is retrieved from Chainlink decentralized
+/// price feeds. Just copy the contract addresses for the symbol that you would like to track from:
+/// https://data.chain.link/feeds.
+///
+///
+/// Note: Rustlink is designed to be ran on the main thread, in an asynchronous tokio environment
+#[derive(Clone)]
+pub struct CryptoPrices {
+    configuration: Configuration,
+    tree: Tree,
+}
+
+#[derive(Error, Debug)]
+pub enum DatabaseError {
+    #[error("Could not find symbol")]
+    NotFound,
+    #[error("Could not deserialize binary data")]
+    Deserialize,
+    #[error("Database internal error: {0}")]
+    SledError(#[from] sled::Error),
+}
+
+impl CryptoPrices {
+    /// Creates a new Rustlink instance.
+    ///
+    /// Expected parameters:
+    /// - `rpc_url`: The RPC url of your chosen EVM network where Chainlink offers decentralised data feeds.
+    /// Don't know where? Check https://data.chain.link/feeds.
+    /// - `fetch_interval_seconds`: How often to update prices in the database (to prevent RPC rate limitation)
+    /// - `sled_path`: The path for the database
+    /// - `contracts`: A tuple list containing a ticker name and its corresponding contract address on the
+    /// EVM chain.
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// use cryptoprices::CryptoPrices;
+    /// let mut contracts:Vec<(String,String)>=Vec::new();
+    /// contracts.push(("BTC".to_string(),"0x9ef1B8c0E4F7dc8bF5719Ea496883DC6401d5b2e".to_string()));
+    /// let crypto_prices=CryptoPrices::new("https://bsc-dataseed1.binance.org/",10,"./cryptoprices",contracts);
+    /// ```
+    pub fn new(
+        rpc_url: &str,
+        fetch_interval_seconds: u64,
+        sled_path: &str,
+        contracts: Vec<(String, String)>,
+    ) -> CryptoPrices {
```

Comment:
Change `CryptoPrices` to `Self`

Perhaps it would make sense to change `new() -> Self` to `try_new() -> Result<Self>` and handle `sled` opening as `let db = sled::open(sled_path)?;` to cascade the error to the invocator?